src/toplevel.c

```diff
@@ -786,31 +797,35 @@ jl_datatype_t *jl_first_argument_datatype(jl_value_t *argtypes)
 }
 
 extern tracer_cb jl_newmeth_tracer;
-JL_DLLEXPORT void jl_method_def(jl_svec_t *argdata, jl_lambda_info_t *f, jl_value_t *isstaged)
+JL_DLLEXPORT void jl_method_def(jl_svec_t *argdata,
+                                jl_source_info_t *f,
+                                jl_value_t *isstaged)
 {
-    // argdata is svec({types...}, svec(typevars...))
-    jl_tupletype_t *argtypes = (jl_tupletype_t*)jl_svecref(argdata,0);
-    jl_svec_t *tvars = (jl_svec_t*)jl_svecref(argdata,1);
+    // argdata is svec(svec(types...), svec(typevars...))
+    jl_svec_t *atypes = (jl_svec_t*)jl_svecref(argdata, 0);
+    jl_svec_t *tvars = (jl_svec_t*)jl_svecref(argdata, 1);
+    size_t nargs = jl_svec_len(atypes);
+    int isva = jl_is_vararg_type(jl_svecref(atypes, nargs - 1));
+    assert(jl_is_svec(atypes));
+    assert(nargs > 0);
+    assert(jl_is_svec(tvars));
+    if (!jl_is_type(jl_svecref(atypes, 0)) || (isva && nargs == 1))
+        jl_error("function type in method definition is not a type");
     jl_methtable_t *mt;
     jl_sym_t *name;
     jl_method_t *m = NULL;
-    JL_GC_PUSH2(&f, &m);
+    jl_tupletype_t *argtype = jl_apply_tuple_type(atypes);
```

Comment:
Ok that's true, but long term, this needs to be a type.

Also: do we want `f(x::T)` to be varargs based on whether `T` evaluates to `Vararg{...}`? That seems like a misfeature to me. And since as you point out `Tuple{Int,Int}` and `Tuple{Vararg{Int,2}}` are equal types, the presence of `Vararg` is not a good basis for identifying varargs functions. It seems like this needs to be syntax-based and passed as a separate flag, since it's not really a property of the signature but of how argument names are bound to argument values. I suppose we can deal with that later though.